src/Microsoft.Health.Fhir.Core/Features/Validation/NarrativeValidator.cs

```diff
@@ -31,19 +30,17 @@ public override IEnumerable<ValidationFailure> Validate(PropertyValidatorContext
 
             if (context.PropertyValue is ResourceElement resourceElement)
             {
-                var poco = resourceElement.ToPoco();
-
-                if (poco is DomainResource resource)
+                if (resourceElement.IsDomainResource())
                 {
-                    foreach (ValidationFailure validationFailure in ValidateResource(resource))
+                    foreach (ValidationFailure validationFailure in ValidateResource(resourceElement))
                     {
                         yield return validationFailure;
                     }
                 }
-
-                if (poco is Bundle bundle)
+                else if (resourceElement.InstanceType.Equals("Bundle", System.StringComparison.OrdinalIgnoreCase))
                 {
-                    var domainResources = bundle.Entry.Select(x => x.Resource).OfType<DomainResource>();
+                    var domainResources = resourceElement.Instance.Select("Bundle.entry.resource")
+                        .Select(e => new ResourceElement(e)).Where(r => r.IsDomainResource());
```

Comment:
Is there a case there `resourceElement.Instance.Select("Bundle.entry.resource")` might return null?

Reply:
I think we've also defined `Bundle.entry.resource` a few times now //cc @brandonpollett